c++ training/LinkedLists/LinkedLists/SinglyLinkedList.h

```diff
@@ -0,0 +1,315 @@
+	void Insert(const T& Value, const int Index)
+	{
+		if (Index < 0 || Index > Size)
+		{
+			return;
+		}
+
+		if (Index == 0)
+		{
+			AddHead(Value);
+			return;
+		}
+		else
+		{
+			FNode& PreviousNode = GetNodeAt(Index - 1);
+			PreviousNode.Next = new FNode(Value, PreviousNode.Next);
+			
+			if (Index == Size)
+			{
+				Tail = Tail->Next;
+			}
+
+			Size++;
+		}
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	void Remove(const int Index)
+	{
+		if (Index < 0 || Index >= Size)
+		{
+			return;
+		}
+
+		if (Index == 0)
+		{
+			FNode* NodeToDeletePtr = Head;
+			Head = NodeToDeletePtr->Next;
+
+			delete NodeToDeletePtr;
+			NodeToDeletePtr = nullptr;
+
+			if (Size == 1)
+			{
+				Tail = nullptr;
+			}
+		}
+		else
+		{
+			FNode& PreviousNode = GetNodeAt(Index - 1);
+			FNode* NodeToDeletePtr = PreviousNode.Next;
+
+			PreviousNode.Next = NodeToDeletePtr->Next;
+
+			delete NodeToDeletePtr;
+			NodeToDeletePtr = nullptr;
+
+			if (Index == Size - 1)
+			{
+				Tail = &PreviousNode;
+			}
+		}
+
+		Size--;
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	void Clear()
+	{
+		int OriginalSize = Size;
+		for (int i = 0; i < OriginalSize; i++)
+		{
+			Remove(0);
+		}
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	class FIterator
+	{
+	private:
+		FNode* Current;
+
+	public:
+		FIterator(FNode* NewCurrent) : Current(NewCurrent) {}
+
+		FIterator& operator++()
+		{
+			Current = Current->Next;
+			return *this;
+		}
+
+		T& operator*()
+		{
+			return Current->Value;
+		}
+
+		bool operator!=(const FIterator& Other)
+		{
+			return Current != Other.Current;
+		}
+	};
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	FIterator begin() const
+	{
+		return FIterator(Head);
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	FIterator end() const
+	{
+		return FIterator(nullptr);
+	}
+	//---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+	template<typename Pred>
+	void ForEach(const Pred& Predicate)
+	{
+		for (T& Item : *this)
+		{
```

Comment:
Wow, I didn't have any idea that I could do something like this, nice

Reply:
like *this v;;